server/common/config/app_config.py

```diff
@@ -0,0 +1,172 @@
+import yaml
+from flatten_dict import unflatten
+import os
+from server.common.config.dataset_config import DatasetConfig
+from server.common.config.server_config import ServerConfig
+from server.common.errors import ConfigurationError
+
+
+class AppConfig(object):
+    """AppConfig stores all the configuration for cellxgene.  The configuration is divided into two main parts:
+    server attributes, and dataset attributes.
+    The server_config contains attributes that refer to the server process as a whole.
+    The default_dataset_config refers to attributes that are associated with the features and
+    presentations of a dataset.
+    The dataset config attributes can be overridden depending on the url by which the
+    dataset was accessed.  These are stored in dataroot_config.
+    AppConfig has methods to initialize, modify, and access the configuration.
+    """
+
+    def __init__(self):
+
+        # the default configuration (see default_config.py)
+        # TODO @madison -- if we always read from the default config (hard coded path) can we set those values as
```

Comment:
I like this idea - @bmccandless is there a reason why the hardcoded default config values can't be set as default values in code?

Reply:
created a ticket --  https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/cellxgene/1880

Reply:
That ideas fine if we remove the default config file, in fact it would be basically required.

But I'm unsure what the need or benefit is of removing the config file
The config file servers a few purposes.
1.  a file to document the attributes
2. a file to see the whole structure in one place 
3. a file to provide an order for the attributes
5. a file, whose contents we can present when "cellxgene launch --dump-default-config" (related to 3).
4. a file that is versioned, so you can easily see what changes PRs

We can find workarounds for all these, but that would just move complexity to other places.  
I don't think there is a good return on investment.

If we want to keep improving the app config, 
here are a few items that I think have more value:
1.  better handling for aws secrets 
2. consolidate all the single_dataset testing that in spread around the app config code into one place
3. better documentation for the config parameters

Reply:
two more idea for improvement

4.  The ability to check that a config file is valid: e.g. `cellxgene launch --check-config config.yaml`
5. The ability to pass in more than one config file.  This would be super useful for the hosted cellxgene deployments.

I think @MDunitz has made an issue for the 4th one.  I can make issues for the others.

Reply:
I think 2 and 3 would be really helpful. I've already created a ticket for 3 https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/cellxgene/1879